### PR TITLE
Incorrect constrained virtual call in method using gsharedvt for reference type.

### PR DESCRIPTION
**Pull request by lateralusX**
For generic constrained virtual calls using reference type's, the virtual call will be dispatched using instance vtable. When the same virtual call is taking place in a gsharedvt method, mono_gsharedvt_constrained_call gets called that will assist making the call, the problem is that it will
use the constrained class for reference types meaning that it will not call the most derived implementation of the method. This difference in the behavior compared to none gsharedvt versions of methods, gives inconsistent results.

One way this will reproduce is by using Dictionary<Type, EmptyStruct>. For methods that will take both Type (reference type) and EmptyStruct (value type) like Add, calling that through IDictionary (hiding details from compiler) will use the gsharedvt version of the method. Internally that method will call Type.GetHashCode (), and since Type really is a RuntimeType it overrides GetHashCode. But since mono_gsharedvt_constrained_call will use the constrained class instead of the instance class, the called method will be Object.GetHashCode, that gives a different result. When later ContainsKey is called using the same Type, it only takes a reference value, so not using a gsharedvt version of the method, then the most derived version of GetHashCode will be called, given a different hash, not finding item in dictionary.

Fix is to make sure behavior in mono_gsharedvt_constrained_call match behavior for regular constrained virtual calls when having a reference type. This makes sure we always call the most  derived version of the method.

Commit also adds a test for this specific scenario.

### CODE REVIEW

File: mono/mini/gshared.cs [hunk 1]
```diff
@@ -1058,6 +1058,14 @@ public enum AnEnum {
 		return 0;
 	}
 
+	public static int test_0_contains_key_constrained_get_hash () {
+		IDictionary<Type, EmptyStruct> dict = new Dictionary<Type, EmptyStruct> ();
```
Comment (vargaz) [inline mono/mini/gshared.cs]: Would be nice to have a lower level test.

Comment (vargaz) [reply mono/mini/gshared.cs]: Made one:
https://gist.github.com/vargaz/a55dbc9cafefd6e8a34d7575872363c1

Comment (lateralusX) [reply mono/mini/gshared.cs]: Works, switched test.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Pull request merged**